File: go/library/mux.go
package library

import jpl "github.com/jplorg/jpl/go/v2/jpl"

// Multiplex the specified array of arguments by calling cb for all possible combinations of arguments.
//
// `mux([[1,2], [3,4]], cb)` for example yields:
// - `cb(1, 3)`
// - `cb(1, 4)`
// - `cb(2, 3)`
// - `cb(2, 4)`
func Mux[Input any](args [][]Input, cb jpl.IMuxer[Input]) jpl.JPLError {
	argCount := len(args)
	if argCount == 1 {
		for _, arg := range args[0] {
			if err := cb.Mux(arg); err != nil {
				return err
			}
		}
		return nil
	}
	execCount := 1
	indices := make([]*iteratorIndex[Input], argCount)
	for i, arg := range args {
		argLength := len(arg)
		execCount *= argLength
		indices[i] = &iteratorIndex[Input]{argLength, 0, arg}
	}
	if execCount == 0 {
		return nil
	}
	buffer := make([]Input, argCount)
	for i, arg := range args {
		buffer[i] = arg[0]
	}
	execIndex := 0
	for {
		if err := cb.Mux(buffer...); err != nil {
			return err
		}
		execIndex += 1
		if execIndex >= execCount {
			break
		}
		// determine next combination
		for argIndex := argCount - 1; argIndex >= 0; argIndex -= 1 {
			arg := indices[argIndex]
			next := arg.current + 1
			if next < arg.max {
				arg.current = next
				buffer[argIndex] = arg.values[next]
				break
			}
			arg.current = 0
			buffer[argIndex] = arg.values[0]
		}
	}
	return nil
}

// Multiplex the specified array of arguments and return the results produced by the callbacks
func MuxOne[Input any, Output any](args [][]Input, cb jpl.IOMuxer[Input, Output]) ([]Output, jpl.JPLError) {
	argCount := len(args)
	if argCount == 1 {
		inputs := args[0]
		result := make([]Output, len(inputs))
		for i, arg := range inputs {
			var err jpl.JPLError
			if result[i], err = cb.Mux(arg); err != nil {
				return nil, err
			}
		}
		return result, nil
	}
	execCount := 1
	indices := make([]*iteratorIndex[Input], argCount)
	for i, arg := range args {
		argLength := len(arg)
		execCount *= argLength
		indices[i] = &iteratorIndex[Input]{argLength, 0, arg}
	}
	if execCount == 0 {
		return nil, nil
	}
	outputs := make([]Output, execCount)
	buffer := make([]Input, argCount)
	for i, arg := range args {
		buffer[i] = arg[0]
	}
	execIndex := 0
	for {
		var err jpl.JPLError
		if outputs[execIndex], err = cb.Mux(buffer...); err != nil {
			return nil, err
		}
		execIndex += 1
		if execIndex >= execCount {
			break
		}
		// determine next combination
		for argIndex := argCount - 1; argIndex >= 0; argIndex -= 1 {
			arg := indices[argIndex]
			next := arg.current + 1
			if next < arg.max {
				arg.current = next
				buffer[argIndex] = arg.values[next]
				break
			}
			arg.current = 0
			buffer[argIndex] = arg.values[0]
		}
	}
	return outputs, nil
}

// Multiplex the specified array of arguments and return a single array of all merged result arrays produced by the callbacks
func MuxAll[Input any, Output any](args [][]Input, cb jpl.IOMuxer[Input, []Output]) ([]Output, jpl.JPLError) {
	segments, err := MuxOne(args, cb)
	if err != nil {
		return nil, err
	}
	return MergeSegments(segments), nil
}

// Create a single array from the specified array segments
func MergeSegments[Value any](segments [][]Value) []Value {
	if len(segments) == 0 {
		return nil
	}
	if len(segments) == 1 {
		return segments[0]
	}
	var count int
	for _, segment := range segments {
		count += len(segment)
	}
	result := make([]Value, 0, count)
	for _, segment := range segments {
		result = append(result, segment...)
	}
	return result
}

File: js/src/library/mux.js
/**
 * Multiplex the specified array of arguments by calling cb for all possible combinations of arguments.
 * If the specified `args` array is empty, `cb` is called once without any arguments.
 *
 * `mux([[1, 2], [3, 4]], cb)` for example calls:
 * - `cb(1, 3)`
 * - `cb(1, 4)`
 * - `cb(2, 3)`
 * - `cb(2, 4)`
 */
function mux(args, cb) {
  const argCount = args.length;
  if (argCount === 1) {
    args[0].forEach((arg) => cb(arg));
    return;
  }
  let execCount = 1;
  const indices = args.map((arg) => {
    const argLength = arg?.length ?? 0;
    execCount *= argLength;
    return [argLength, 0, arg];
  });
  if (execCount === 0) return;
  const buffer = args.map(([v]) => v);
  let execIndex = 0;
  for (;;) {
    cb(...buffer);
    execIndex += 1;
    if (execIndex >= execCount) break;
    // determine next combination
    for (let argIndex = argCount - 1; argIndex >= 0; argIndex -= 1) {
      const arg = indices[argIndex];
      const [max, current, values] = arg;
      const next = current + 1;
      if (next < max) {
        arg[1] = next;
        buffer[argIndex] = values[next];
        break;
      }
      arg[1] = 0;
      [buffer[argIndex]] = values;
    }
  }
}

export default mux;

/** Multiplex the specified array of arguments and return the results produced by the callbacks */
export function muxOne(args, cb) {
  const argCount = args.length;
  if (argCount === 1) return args[0].map((arg) => cb(arg));
  let execCount = 1;
  const indices = args.map((arg) => {
    const argLength = arg?.length ?? 0;
    execCount *= argLength;
    return [argLength, 0, arg];
  });
  if (execCount === 0) return [];
  const outputs = new Array(execCount);
  const buffer = args.map(([v]) => v);
  let execIndex = 0;
  for (;;) {
    outputs[execIndex] = cb(...buffer);
    execIndex += 1;
    if (execIndex >= execCount) break;
    // determine next combination
    for (let argIndex = argCount - 1; argIndex >= 0; argIndex -= 1) {
      const arg = indices[argIndex];
      const [max, current, values] = arg;
      const next = current + 1;
      if (next < max) {
        arg[1] = next;
        buffer[argIndex] = values[next];
        break;
      }
      arg[1] = 0;
      [buffer[argIndex]] = values;
    }
  }
  return outputs;
}

/** Multiplex the specified array of arguments asynchronously and return the results produced by the callbacks */
export function muxAsync(args, cb) {
  return Promise.all(muxOne(args, cb));
}

/** Multiplex the specified array of arguments asynchronously and return a single array of all merged result arrays produced by the callbacks */
export async function muxAll(args, cb) {
  return mergeSegments(await muxAsync(args, cb));
}

/** Create a single array from the specified array segments */
export function mergeSegments(segments) {
  if (segments.length === 0) return segments;
  if (segments.length === 1) return segments[0];
  return segments.flat(1);
}
